Fix #2906: warpPedIntoVehicle causes a C++ runtime assertion failure

TracerDS opened the pull request:
`GetMaxPassengers()` returns `VEHICLE_PASSENGERS_UNDEFINED` if the amount of seat is undefined (this happens mainly on trailers). We can't warp player to a vehicle that doesn't have seats.
By adding this fix, server wont crash from runtime exception when warping to a vehicle that doesnt have "enough" seats

<hr/>

In #2918 I fixed the issue but my repo and branch got corrupted and broke the whole PR (couldnt change the commit, add or remove files). Sorry

TracerDS committed:
Fixed issue #2906

lopezloo reviewed (dismissed):
Your change disables warping to trailers at driver seat too.

How about also adding `seat < MAX_VEHICLE_SEATS` checks to `CVehicle::SetOccupant` and `CPed::SetOccupiedVehicle`?

TracerDS commented:
That would be a good idea, I'll check it
Although trailers shouldnt be driveable anyway

lopezloo commented:
Or maybe just only to `CVehicle::SetOccupant()` though as `MAX_VEHICLE_SEATS` is vehicle related definition. Not sure if adding it also to `CPed::SetOccupiedVehicle()` would be good coding practice.

TracerDS committed:
#2920 - Added checks for vehicle occupants

lopezloo commented on Server/mods/deathmatch/logic/CVehicle.cpp:
```diff
@@ -644,7 +647,7 @@ bool CVehicle::SetOccupant(CPed* pPed, unsigned int uiSeat)
     }
 
-    return true;
+    return true; // return true? even if function didn't end successfully?
```
It ends successfully. That `bAlreadySetting` act as lock to prevent infinite recursion.

lopezloo commented on Server/mods/deathmatch/logic/CVehicle.cpp:
```diff
@@ -618,6 +618,9 @@ bool CVehicle::SetOccupant(CPed* pPed, unsigned int uiSeat)
     static bool bAlreadySetting = false;
     if (!bAlreadySetting)
     {
+        if (uiSeat < MAX_VEHICLE_SEATS)
```
```suggestion
        if (uiSeat >= MAX_VEHICLE_SEATS)
```

TracerDS committed:
Update Server/mods/deathmatch/logic/CVehicle.cpp

Damn, i mismatched the sign,

Co-authored-by: lopez <7338099+lopezloo@users.noreply.github.com>

lopezloo commented on Server/mods/deathmatch/logic/CPed.cpp:
```diff
@@ -406,6 +406,9 @@ CVehicle* CPed::SetOccupiedVehicle(CVehicle* pVehicle, unsigned int uiSeat)
     {
+        if (uiSeat < MAX_VEHICLE_SEATS)
+            return m_pVehicle;
+
```
Maybe drop that check altogether from there. Looks a bit weird to me.

Also, why that method even return `CVehicle`... If anything, it could return boolean. But probably refactoring it is a job for seperate PR.

TracerDS replied:
Checking from the vehicle's side only then?

lopezloo replied:
Yes.

TracerDS committed:
#2920 - Review change

Removed seat check in CPed.cpp

lopezloo commented on Server/mods/deathmatch/logic/CStaticFunctionDefinitions.cpp:
```diff
@@ -4136,6 +4136,9 @@ bool CStaticFunctionDefinitions::WarpPedIntoVehicle(CPed* pPed, CVehicle* pVehic
     assert(pPed);
     assert(pVehicle);
 
+    if (pVehicle->GetMaxPassengers() == VEHICLE_PASSENGERS_UNDEFINED)
```
```suggestion
    if (uiSeat > 0 && pVehicle->GetMaxPassengers() == VEHICLE_PASSENGERS_UNDEFINED)
```

TracerDS replied:
~~There is already a line regarding the `uiSeat` variable (`if (uiSeat <= pVehicle->GetMaxPassengers() &&...`).
That change might be pointless~~

Nvm that check isnt quite enough

lopezloo replied:
By checking if seat is higher than 0 here we can allow warping to vehicles on driver seat even if passenger seat count is undefined.

TracerDS replied:
But should you drive a trailer at all? They should be attached to other vehicle but not driven directly

TracerDS committed:
Reformatted CStaticFunctionDefinitions

Changed seat check to include values less than 0 and bigger than max passengers of a vehicle

lopezloo commented:
It's still not possible to warp player to driver seat in trailer vehicles like it was before. Excluding seat 0 in `pVehicle->GetMaxPassengers() == VEHICLE_PASSENGERS_UNDEFINED` check would fix that.

TracerDS committed:
#2920 - Review change

Reformatted seat check.
Now its possible to enter vehicle as a driver on trailers

TracerDS committed:
Merge branch 'master' into mtasa-updates

TracerDS committed:
Merge branch 'master' into mtasa-updates

lopezloo commented on Server/mods/deathmatch/logic/CStaticFunctionDefinitions.cpp:
```diff
@@ -4136,10 +4136,16 @@ bool CStaticFunctionDefinitions::WarpPedIntoVehicle(CPed* pPed, CVehicle* pVehic
+    if (uiSeat > pVehicle->GetMaxPassengers() ||
+        (uiSeat > 0 && pVehicle->GetMaxPassengers() == VEHICLE_PASSENGERS_UNDEFINED)
+    )
+        return false;
```
I think 2 seperate if statements would be more readable.
```suggestion
    if (uiSeat > pVehicle->GetMaxPassengers())
        return false;

    if (uiSeat > 0 && pVehicle->GetMaxPassengers() == VEHICLE_PASSENGERS_UNDEFINED)
        return false
```

TracerDS committed:
Update Server/mods/deathmatch/logic/CStaticFunctionDefinitions.cpp

Increased readability

Co-authored-by: lopez <7338099+lopezloo@users.noreply.github.com>

TracerDS commented on Server/mods/deathmatch/logic/CStaticFunctionDefinitions.cpp:
```diff
@@ -4136,10 +4136,17 @@ bool CStaticFunctionDefinitions::WarpPedIntoVehicle(CPed* pPed, CVehicle* pVehic
+        return false;
+
+    if (uiSeat > 0 && pVehicle->GetMaxPassengers() == VEHICLE_PASSENGERS_UNDEFINED)
+        return false
```
Forgot about `;` after `return false`. Whoopsie

TracerDS committed:
Update CStaticFunctionDefinitions.cpp

Fixed a typo

patrikjuvonen committed:
Merge branch 'master' into mtasa-updates

lopezloo committed:
Merge branch 'master' into mtasa-updates

lopezloo committed:
Remove empty line

lopezloo reviewed (approved):
Looks good to me. Thank you.
